extensions/haskell_toolchains.bzl

```diff
@@ -0,0 +1,165 @@
+        ),
+    },
+    doc = """See [rules_haskell_toolchains](toolchain.html#rules_haskell_toolchains).
+    Only the first `bindists` tag is taken into account (according to the iteration order over modules).
+    """,
+)
+
+_bindist_tag = tag_class(
+    attrs = {
+        "name": attr.string(
+            mandatory = True,
+            doc = "A unique name for the repository",
+        ),
+        "version": attr.string(
+            mandatory = True,
+            doc = "The desired GHC version",
+        ),
+        "target": attr.string(
+            mandatory = True,
+            doc = "The desired architecture (See [ghc_bindist_generated.bzl](https://github.com/tweag/rules_haskell/blob/master/haskell/private/ghc_bindist_generated.bzl))",
+        ),
+        "ghcopts": attr.string_list(
+            doc = "[see rules_haskell_toolchains](toolchain.html#rules_haskell_toolchains-ghcopts)",
+        ),
+        "haddock_flags": attr.string_list(
+            doc = "[see rules_haskell_toolchains](toolchain.html#rules_haskell_toolchains-haddock_flags)",
+        ),
+        "repl_ghci_args": attr.string_list(
+            doc = "[see rules_haskell_toolchains](toolchain.html#rules_haskell_toolchains-repl_ghci_args) ",
+        ),
+        "cabalopts": attr.string_list(
+            doc = "[see rules_haskell_toolchains](toolchain.html#rules_haskell_toolchains-cabalopts)",
+        ),
+        "locale": attr.string(
+            doc = "[see rules_haskell_toolchains](toolchain.html#rules_haskell_toolchains-locale)",
+        ),
+    },
+    doc = "Declares and configure a bindist haskell toolchain. See [ghc_bindist](ghc_bindist.html#ghc_bindist).",
+)
+
+def _all_toolchains_impl(rctx):
+    content = "\n".join(rctx.attr.toolchains)
+    rctx.file("BUILD.bazel", content = content)
+
+_all_toolchains = repository_rule(
+    implementation = _all_toolchains_impl,
+    attrs = {
+        "toolchains": attr.string_list(
+            doc = "All the `toolchain(...)` declarations for the bindist toolchains as string",
+        ),
+    },
+    doc = """ Used to generate the `all_bindist_toolchains` external repository.
+    We can then invoque `register_toolchains("@all_bindist_toolchains//:all")` in the MODULE.bazel file.
+    """,
+)
+
+def _haskell_toolchains_impl(mctx):
+    # We gather the declarations of all the toolchains in the `toolchain_declarations` list
+    # in order to write them to the `@all_bindist_toolchains` repository.
+
+    # The code could be simplified a bit by registering aliases to the
+    # `toolchain` rules, once the following issue is resolved
+    # https://github.com/bazelbuild/bazel/issues/16298
+    toolchain_declarations = []
+
+    toolchain_names = []
+    found_bindists = False
+    for module in mctx.modules:
+        for bindist_tag in module.tags.bindist:
+            name = "bindist_{}_{}_{}".format(module.name, module.version, bindist_tag.name)
```

Comment:
Is `bindist_tag.name` actually needed, or could it be replaced by an autogenerated unique suffix (e.g. a counter, or perhaps [a `hash`](https://bazel.build/rules/lib/globals/all#hash) plus collision counter to reduce invalidation)? Asking in the spirit of [this discussion](https://github.com/bazel-contrib/rules-template/pull/46#discussion_r1097502233). Since the name is scoped to within the calling MODULE and not global it doesn't seem too onerous, but, maybe it can be avoided altogether.

Reply:
Ah indeed I think we can do without the name attribute (and without auto-generated suffixes).

I updated the code to rely on the `target` attribute instead.

Before, if two `bindist` tags shared the same `target` attribute, the toolchains would have the same constraints so the second one would be effectively ignored. So now, a module is not allowed to use two `bindist` tags with the same `target`, and `(module.name, module.version, bindist_tag.target)` can be used as an identifier.